### PR TITLE
[parser] enh: allow whitespace in Format specification

### PR DESCRIPTION
**Pull request by gxyd**
Blank characters may precede the initial left parenthesis of the format specification and additional blank characters may appear at any point within the format specification, with no effect on the interpretation of the format specification.

e.g.; `(I 1 2)` is same as `(I12)` for format specification and are both valid.

Earlier, we used to raise tokenizer error because of no support for it

issue was initially raised in #4040

### CODE REVIEW

File: tests/format1.f90 [hunk 1]
```diff
@@ -1,60 +1,112 @@
 program format_04
+    real :: a,b,c,d,e(6)
```
Comment (gxyd) [inline tests/format1.f90:2]: the general principal with the test cases is, for each of the test-case present earlier in `tests/format1.f90`, I've added another similar test-case but using `LABEL FORMAT ...` to use for printing purpose instead of character string representation of format specification and I've made sure to introduce whitespaces at the necessary places in `xxx` part of `LABEL FORMAT xxxx` to ensure that the changes work just well.

File: tests/format1.f90 [hunk 2]
```diff
@@ -1,60 +1,112 @@
...
-print "(F10.3)", t2-t1
-print "(F0.6)", real_hundred
-
-! test for issue: https://github.com/lfortran/lfortran/issues/4040
-print "(2 (I3))", i, j
-print "(2 (I 3))", i, j
-print 9, i, i + 1, j + 1, i + 2, j + 2
-! the below test also ensures that blank character
-! isn't removed from ' Dates: '
-9 FORMAT (I12, /, ' Dates: ', 2 (2I3, I5))
```
Comment (certik) [inline tests/format1.f90:58]: This one with `I12` doesn't seem to be tested anymore.

**Comment by certik** [comment]
The failure is:
```
format1.f90 * run
The JSON metadata differs against reference results
Reference JSON: tests/reference/run-format1-04216cf.json
Output JSON:    tests/output/run-format1-04216cf.json
Omitting 10 identical items
Differing items:
{'stdout_hash': '3ef6512ee97acc29ff3be42302f39e69449b6e7ce94f096536ce9436'} != {'stdout_hash': '3ffedee945d8f97620144a79854547c60ad12af5f2e2bde79f28eeec'}
Diff against: tests/reference/run-format1-04216cf.stdout
68c68
<      0.000
---
>    100.000
```

File: tests/format1.f90 (hunk 2 above)
Comment (gxyd) [reply tests/format1.f90:58]: I actually made it as `I 12` (i.e. with a space) here: https://github.com/lfortran/lfortran/pull/4060/files#diff-fe371b65a59e57bc5e5a44746ef8b648ca4c064435f3bcf83d64781342d96799R109.

I can change it back to `I12` if we like that as well.

Comment (gxyd) [reply tests/format1.f90:58]: or another test maybe.

Comment (certik) [reply tests/format1.f90:58]: We need to test both `I12` and `I 12`. So leave the original test, and add a new one with the space.

File: tests/format1.f90 [hunk 3]
```diff
@@ -99,14 +99,16 @@ program format_04
     ! test for issue: https://github.com/lfortran/lfortran/issues/4040
     print "(2 (I3))", i, j
     print 25, i, j
-    25 FORMAT (F10.3)
```
Comment (gxyd) [inline tests/format1.f90]: this was the reason for failure of CI earlier, I added a wrong test for `(F10.3)` for integer values (i.e. `i, j`). We should raise an error anyway (that's what GFortran does as well) instead of passing without error (which we currently do).

**Comment by gxyd** [comment]
I think the current failure of CI isn't related to this PR.

File: tests/format1.f90 (hunk 2 above)
Comment (gxyd) [reply tests/format1.f90:58]: Done, I've kept the already present test and added another as well.

**Comment by certik** [comment]
The failure is due to https://github.com/lfortran/lfortran/issues/3985.

File: src/lfortran/parser/tokenizer.re [hunk 4]
```diff
@@ -753,31 +753,32 @@ void lex_format(unsigned char *&cur, Location &loc,
...
+
             position_edit_desc
-                = 'T' int
-                | 'TL' int
-                | 'TR' int
-                | int 'X'
+                = whitespace? 'T' whitespace? int
+                | whitespace? 'T' whitespace? 'L' whitespace? int
+                | whitespace 'T' whitespace? 'R' whitespace? int
+                | whitespace? int whitespace? 'X'
```
Comment (certik) [inline src/lfortran/parser/tokenizer.re:781]: Now when tests pass, let's fix this. In the main parser we simply skip whitespace. However, here we need to skip whitespace in the middle of a rule. I think the `whitespace?` at the beginning can be taken care of using that approach, but most of the `whitespace?` couldn't.

Another approach is to remove the white space in a prescanner, before tokenizing it. We do that for the fixed-form. But in the free-form case, it would slow down compilation.

So the above approach is currently the only one that I can think of, but I don't like how verbose it is.

**Review by certik — APPROVED** [review]

File: src/lfortran/parser/tokenizer.re (hunk 4 above)
Comment (gxyd) [reply src/lfortran/parser/tokenizer.re]: Could you take a look at the latest commit that I pushed maybe?

**Comment by gxyd** [comment]
> The failure is due to https://github.com/lfortran/lfortran/issues/3985.

I'll give that a look sometime I think.

File: tests/format1.f90 (hunk 2 above)
Comment (certik) [reply tests/format1.f90:58]: Previously we had a test for `I12`, but now when I search `I12` in the new file, it's not there, so I think we are not testing it anymore. Can we test the original test and then in addition add more tests for the spaces?

**Comment by certik** [comment]
I think it's fine. Not as verbose as before. I can't think right now of any way to simplify, so let's go with this.

Let's restore the original test and then we can merge it.

File: tests/format1.f90 (hunk 2 above)
Comment (gxyd) [reply tests/format1.f90:58]: Alright, I've made sure now that `I12` test case is definitely there, and I've added more test cases to ensure that format description with whitespace works in all cases.
